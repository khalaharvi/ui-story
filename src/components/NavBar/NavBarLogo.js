import React from 'react';
import PropTypes from 'prop-types';

import { createClassNames } from 'utils';
import Link from 'components/Link';
import Image from 'components/Image';

const cx = createClassNames('navbar-logo');

// TO DO: allow for custom width/height

const NavBarLogo = ({
  alt,
  logo,
  src,
  to
}) => {
  const imgOrElement = typeof logo !== 'undefined' ? logo : <Image src={src} alt={alt} />;

  return (
    <div className={cx('')}>
      {to ? <Link to={to}>{imgOrElement}</Link> : imgOrElement}
    </div>
  );
};

NavBarLogo.propTypes = {
  alt: (props, propName) => {
    const alt = props[propName];

    if (typeof props.logo === 'undefined' && typeof alt !== 'string') {
      throw new Error("'alt' is required.");
    }
  },
  logo: (props, propName) => {
    const logo = props[propName];

    if (typeof props.src === 'undefined' && !React.isValidElement(logo)) {
      throw new Error("'logo' is required and must be a valid React element.");
    }
  },
  src: (props, propName) => {
    const src = props[propName];

    if (typeof props.logo === 'undefined' && typeof src !== 'string') {
      throw new Error("'src' is required.");
    }
  },
  to: PropTypes.string
};

export default NavBarLogo;
