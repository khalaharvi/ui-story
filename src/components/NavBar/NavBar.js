import React, { Component, Children } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { createClassNames } from 'utils';
import {
  Search as SearchIcon,
  Reset as ResetIcon
} from 'icons';
import PoweredBy from 'components/PoweredBy';
import NavBarLogo from './NavBarLogo';

const cx = createClassNames('navbar');

const getNavBarLogoProps = (logo, logoLinkTo) => {
  let navBarLogoProps = {
    to: logoLinkTo || null
  };

  if (React.isValidElement(logo)) {
    navBarLogoProps = Object.assign({}, navBarLogoProps, { logo });
  } else {
    navBarLogoProps = Object.assign({}, navBarLogoProps, {
      src: logo.src,
      alt: logo.alt
    });
  }

  return navBarLogoProps;
};

class NavBar extends Component {
  static propTypes = {
    className: PropTypes.string,
    logo: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.shape({
        src: PropTypes.string.isRequired,
        alt: PropTypes.string.isRequired
      })
    ]),
    logoLinkTo: PropTypes.string,
    searchInputComponent: PropTypes.func.isRequired,
    poweredBy: PropTypes.bool,
    mobile: PropTypes.bool
  };

  static defaultProps = {
    poweredBy: true,
    mobile: false
  };

  state = {
    searchInputVisible: false
  };

  toggleSearchInput = (e) => {
    e.preventDefault();

    this.setState({ searchInputVisible: !this.state.searchInputVisible }, () => {
      if (this.state.searchInputVisible) {
        if (this.searchInput) {
          this.searchInput.focus();
        }
      }
    });
  };

  render() {
    const {
      logo,
      logoLinkTo,
      children,
      searchInputComponent: SearchInputComponent,
      className,
      poweredBy,
      mobile
    } = this.props;

    return (
      <div
        className={classnames(
          cx(''),
          className,
          {
            [cx('mobile')]: mobile
          }
        )}
      >
        <div className={cx('inner')}>
          { logo && <NavBarLogo {...getNavBarLogoProps(logo, logoLinkTo)} />}
          {
            (!mobile || (mobile && !this.state.searchInputVisible)) &&
              <nav className={cx('nav')}>
                <ul className={cx('nav-list')}>
                  {
                    Children.map(children, child => (
                      <li className={cx('nav-list-item')}>{child}</li>
                    ))
                  }
                </ul>
              </nav>
          }
          {
            (!mobile || (mobile && this.state.searchInputVisible)) &&
              <div className={cx('search')}>
                <SearchInputComponent
                  inputRef={(searchInput) => {
                    this.searchInput = searchInput;
                  }}
                />
              </div>
          }
          {(poweredBy && !mobile) && <PoweredBy />}
          {mobile &&
            <button onClick={this.toggleSearchInput} className={cx('search-toggle')}>
              { this.state.searchInputVisible ? <ResetIcon /> : <SearchIcon />}
            </button>
          }
        </div>
      </div>
    );
  }
}

export default NavBar;
