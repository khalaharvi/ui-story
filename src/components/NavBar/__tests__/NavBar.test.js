import React from 'react';
import { shallow } from 'enzyme';

import NavBar from '../NavBar';

const SearchInput = () => (<div>Search Input</div>);

describe('<NavBar />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <NavBar
        logo={{ src: 'https://placehold.it/259x28', alt: 'Logo' }}
        searchInputComponent={() => <SearchInput />}
      >
        Item
      </NavBar>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with className', () => {
    const wrapper = shallow(
      <NavBar
        logo={{ src: 'https://placehold.it/259x28', alt: 'Logo' }}
        searchInputComponent={() => <SearchInput />}
        className="custom-nav-bar"
      >
        Item
      </NavBar>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
