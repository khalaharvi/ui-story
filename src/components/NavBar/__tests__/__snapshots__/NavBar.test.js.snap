// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar /> renders correctly 1`] = `
<div
  className="dh-navbar"
>
  <div
    className="dh-navbar-inner"
  >
    <NavBarLogo
      alt="Logo"
      src="https://placehold.it/259x28"
      to={null}
    />
    <nav
      className="dh-navbar-nav"
    >
      <ul
        className="dh-navbar-nav-list"
      >
        <li
          className="dh-navbar-nav-list-item"
          key=".0"
        >
          Item
        </li>
      </ul>
    </nav>
    <div
      className="dh-navbar-search"
    >
      <searchInputComponent
        inputRef={[Function]}
      />
    </div>
    <PoweredBy
      to="https://deringhall.com?ref=ui-kit"
    />
  </div>
</div>
`;

exports[`<NavBar /> renders correctly with className 1`] = `
<div
  className="dh-navbar custom-nav-bar"
>
  <div
    className="dh-navbar-inner"
  >
    <NavBarLogo
      alt="Logo"
      src="https://placehold.it/259x28"
      to={null}
    />
    <nav
      className="dh-navbar-nav"
    >
      <ul
        className="dh-navbar-nav-list"
      >
        <li
          className="dh-navbar-nav-list-item"
          key=".0"
        >
          Item
        </li>
      </ul>
    </nav>
    <div
      className="dh-navbar-search"
    >
      <searchInputComponent
        inputRef={[Function]}
      />
    </div>
    <PoweredBy
      to="https://deringhall.com?ref=ui-kit"
    />
  </div>
</div>
`;
